### PR TITLE
feature: pipeline parallelism support

### PR DESCRIPTION
**Pull request by yezhem**

### CODE REVIEW

**Comment by yezhem** [comment]
my test result:
image
image

File: mlora/pipeline/function.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                input_args: MultiLoraBatchData):
+        assert isinstance(tensor_data, torch.Tensor)
+
+        msg = PipeMessage(src_=transport.worker_name,
+                          dst_=transport.next_worker_name,
+                          msg_type_=PipeMessageType.ACTIVATIONS,
+                          msg_id_=msg_id,
+                          tensor_data_=tensor_data,
+                          batch_data_=input_args)
+        transport.send_message(msg, False)
```
Comment (merlintang) [inline mlora/pipeline/function.py:26]: any timeout issue? do we need to check this? with try catch?

File: mlora/pipeline/function.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                phony: torch.Tensor,
+                transport: Transport,
+                msg: PipeMessage) -> torch.Tensor:
+        assert msg.msg_type_ == PipeMessageType.ACTIVATIONS
+        assert isinstance(msg.tensor_data_, torch.Tensor)
+
+        ctx.msg_id_ = msg.msg_id_
+        ctx.transport_ = transport
+
+        return msg.tensor_data_ * phony
```
Comment (merlintang) [inline mlora/pipeline/function.py:50]: can you explain why we need to use tensor_data * phony

File: mlora/pipeline/transport.py
Comment (merlintang) [inline mlora/pipeline/transport.py:1]: can we have a test on the transport

File: mlora/pipeline/queue.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+import torch
+import logging
+
+from queue import Queue
+from threading import Thread
+from typing import Optional
+
+
+class DeviceSwapQueue:
```
Comment (merlintang) [inline mlora/pipeline/queue.py:12]: any test on this part? how to measure the pef for this one ?

File: mlora/pipeline/function.py (hunk 1 above)
Comment (yezhem) [reply mlora/pipeline/function.py:26]: the torch RPC framework will retry

File: mlora/pipeline/queue.py (hunk 3 above)
Comment (yezhem) [reply mlora/pipeline/queue.py:12]: only use the nsys to test the perf, hard to add it to ci.

**Pull request merged**